src/condor_tests/test_fto_failure_propagation.py

```diff
@@ -0,0 +1,843 @@
+                "TransferErrorData" : [
+                    {
+                        "ErrorType" : "Specification",
+                        "ErrorCode" : 405,
+                        "ErrorString" : "Failed to create file foo",
+                        "FailedServer" : "todd.chtc.wisc.edu",
+                    },
+                ],
+            },
+            "output",
+            0,
+            0,
+        ),
+    ),
+}
+
+#--------------------------------------------------------------------------
+# List of 'custom submit macro' -> file contents of file encoded ads for debug plugin
+PLUGIN_ADS = {
+    "parameter_output" :
+"""ErrorCode = 32
+ErrorString = "Invalid host name provided"
+PluginLaunched = True
+DeveloperData = [ ConnectionTime=98.3; ]
+""",
+    "resolution_output" :
+"""ErrorCode = 404
+ErrorString = "Failed to locate host"
+FailedName = "perfect.gaming.host"
+FailureType = "Definitive"
+DeveloperData = [ ConnectionAttempts = 10; Trolled = True; ]
+""",
+    "contact_output" :
+"""ErrorCode = 12
+ErrorString = "Failed to contact server"
+FailedServer = "national.secrets.gov"
+""",
+    "authorization_output" :
+"""ErrorCode = 52
+ErrorString = "Detected bad cookies"
+FailedServer = "cookie.monster.org"
+FailureType = "Authentication"
+ShouldRefresh = True
+DeveloperData = [ Note = "Clear cookie cache"; ]
+""",
+    "specification_output" :
+"""ErrorCode = 405
+ErrorString = "Failed to create file foo"
+FailedServer = "todd.chtc.wisc.edu"
+""",
+    "transfer_failure_output" :
+"""ErrorCode = 10
+ErrorString = "Transfer rate fell below threshold: 100 B/s"
+FailedServer = "tj.winblows.net"
+FailureType = "TooSlow"
+""",
+}
+
+
+@action
+def write_plugin_ads(test_dir):
+    """Write all debug plugin ad files for encoding. Return dictionary of 'submit macro' -> 'Absolute path filenames'"""
+    FILES = dict()
+
+    for TAG, AD in PLUGIN_ADS.items():
+        ad_file = test_dir / f"{TAG}.ad"
+        ad_file.write_text(AD)
+        FILES[TAG] = str(ad_file)
+
+    return FILES
+
+#--------------------------------------------------------------------------
+@action
+def submit_jobs(default_condor, path_to_sleep, write_plugin_ads):
+    """Submit all test cases as individual jobs"""
+    HANDLES = dict()
+
+    # Default things to add to all test cases
+    defaults = {
+        "path_to_sleep" : path_to_sleep,  # Custom macro to reference path to sleep
+        "My.TestName" : "$(test_name)",   # Mark test name in job ClassAd
+        "log" : "$(test_name).log",       # Set a log file to the test name
+        "transfer_executable" : False,    # Don't transfer executable (we have absolute path)
+        "should_transfer_files" : True,   # Always transfer files
+    }
+    # Update defaults with macro -> plugin ad files
+    defaults.update(write_plugin_ads)
+
+    # Submit each test job
+    for TEST, details in TEST_CASES.items():
+        desc, _, _ = details
+        desc.update(defaults)
```

Comment:
```suggestion
        desc = {}
        desc.update(defaults)
        desc.update(details)
```
You need to `update()` the test-specific `details` _after_ the `defaults` so that they win.  You can't just assign `defaults` to `desc` to start because that would make `desc` a reference, not a copy, and you don't want to change the defaults from test job to test job.